File: server/src/main/java/com/tp/holdem/server/ConnectedPlayers.java
package com.tp.holdem.server;

import com.tp.holdem.logic.model.PlayerNumber;
import io.vavr.Tuple;
import io.vavr.Tuple2;
import io.vavr.collection.HashMap;
import io.vavr.collection.Map;
import io.vavr.control.Option;
import lombok.AllArgsConstructor;

import java.util.function.BiConsumer;

@AllArgsConstructor
class ConnectedPlayers {
	private final Map<Integer, Tuple2<Boolean, PlayerNumber>> connectedMap;

	static ConnectedPlayers empty() {
		return new ConnectedPlayers(HashMap.empty());
	}

	boolean isConnected(int connectionId) {
		return connectedMap.get(connectionId)
				.exists(Tuple2::_1);
	}

	boolean wasConnected(int connectionId) {
		return connectedMap.containsKey(connectionId);
	}

	Option<Integer> getConnectionId(int playerNumber) {
		return connectedMap
				.mapValues(Tuple2::_2)
				.mapValues(PlayerNumber::getNumber)
				.find(value -> value._2.equals(playerNumber))
				.map(Tuple2::_1);
	}

	Option<Integer> getConnected(int connectionId) {
		return connectedMap.get(connectionId)
				.filter(Tuple2::_1)
				.map(Tuple2::_2)
				.map(PlayerNumber::getNumber);
	}

	ConnectedPlayers connect(Integer connectionId, Integer playerNumber) {
		return new ConnectedPlayers(connectedMap.put(connectionId, Tuple.of(true, PlayerNumber.of(playerNumber))));
	}

	ConnectedPlayers reconnect(Integer connectionId, Integer playerNumber) {
		return new ConnectedPlayers(connectedMap.replaceValue(connectionId, Tuple.of(true, PlayerNumber.of(playerNumber))));
	}

	int size() {
		return connectedMap.size();
	}

	void forEach(BiConsumer<Integer, Integer> action) {
		connectedMap
				.mapValues(Tuple2::_2)
				.mapValues(PlayerNumber::getNumber)
				.forEach(action);
	}
}

File: server/src/main/kotlin/com/tp/holdem/server/ConnectedPlayers.kt
package com.tp.holdem.server

import com.tp.holdem.logic.model.PlayerNumber
import io.vavr.Tuple
import io.vavr.Tuple2
import io.vavr.collection.HashMap
import io.vavr.collection.Map
import io.vavr.control.Option

internal class ConnectedPlayers(private val connectedMap: Map<Int, Tuple2<Boolean, PlayerNumber>>) {
    companion object {
        @JvmStatic
        fun empty(): ConnectedPlayers {
            return ConnectedPlayers(HashMap.empty())
        }
    }

    fun isConnected(connectionId: Int): Boolean {
        return connectedMap.get(connectionId)
                .exists { it._1() }
    }

    fun wasConnected(connectionId: Int): Boolean {
        return connectedMap.containsKey(connectionId)
    }

    fun getConnectionId(playerNumber: PlayerNumber): Option<Int> {
        return if (playerNumber.exists()) getConnectionId(playerNumber.number) else Option.none()
    }

    private fun getConnectionId(playerNumber: Int): Option<Int> {
        return connectedMap
                .mapValues { it._2() }
                .mapValues { it.number }
                .find { value -> value._2 == playerNumber }
                .map { it._1() }
    }

    fun getConnected(connectionId: Int): Option<PlayerNumber> {
        return connectedMap.get(connectionId)
                .filter { it._1() }
                .map { it._2() }
    }

    fun connect(connectionId: Int, playerNumber: PlayerNumber): ConnectedPlayers {
        return ConnectedPlayers(connectedMap.put(connectionId, Tuple.of(true, playerNumber)))
    }

    fun reconnect(connectionId: Int, playerNumber: PlayerNumber): ConnectedPlayers {
        return ConnectedPlayers(connectedMap.replaceValue(connectionId, Tuple.of(true, playerNumber)))
    }

    fun size(): Int {
        return connectedMap.size()
    }

    fun forEach(action: (Int, PlayerNumber) -> Unit) {
        connectedMap
                .mapValues { it._2() }
                .forEach(action)
    }
}
